src/v/s3/signature.cc

```diff
@@ -0,0 +1,399 @@
+///   be added.
+///     For example, if you are using temporary security credentials, you need
+///     to include x-amz-security-token in your request. You must add this
+///     header in the list of CanonicalHeaders.
+///
+///
+static result<canonical_headers>
+get_canonical_headers(const http::client::request_header& request) {
+    std::vector<std::pair<std::string, std::string>> headers;
+    for (const auto& it : request) {
+        auto name = it.name_string();
+        auto value = it.value();
+        headers.emplace_back(name, value);
+    }
+    std::sort(headers.begin(), headers.end());
+    std::string cheaders;
+    std::string snames;
+    int cnt = 0;
+    for (auto [name, value] : headers) {
+        tolower(name);
+        boost::trim(value);
+        cheaders += name;
+        cheaders += ":";
+        cheaders += value;
+        cheaders += "\n";
+        if (cnt++ > 0) {
+            snames.push_back(';');
+        }
+        snames += name;
+    }
+    return canonical_headers{
+      .canonical_headers = cheaders, .signed_headers = snames};
+}
+
+/// Create canonical request:
+/// <HTTPMethod>\n
+/// <CanonicalURI>\n
+/// <CanonicalQueryString>\n
+/// <CanonicalHeaders>\n
+/// <SignedHeaders>\n
+/// <HashedPayload>
+static result<std::string> create_canonical_request(
+  const canonical_headers& hdr,
+  const http::client::request_header& header,
+  const std::string& hashed_payload) {
+    auto method = std::string(header.method_string());
+    auto target = std::string(header.target());
+    if (target.empty() || target.at(0) != '/') {
+        target = "/" + target;
+    }
+    std::stringstream result;
+    auto canonical_uri = get_canonical_uri(target);
+    if (!canonical_uri) {
+        return canonical_uri.error();
+    }
+    auto canonical_query = get_canonical_query_string(target);
+    if (!canonical_query) {
+        return canonical_query.error();
+    }
+    result << method << '\n';
+    result << canonical_uri.value() << '\n';
+    result << canonical_query.value() << '\n';
+    result << hdr.canonical_headers << '\n';
+    result << hdr.signed_headers << '\n';
+    result << hashed_payload;
+    return result.str();
+}
+
+/// Genertes string-to-sign (in spec terms), example:
+///
+///   "AWS4-HMAC-SHA256" + "\n" +
+///   timeStampISO8601Format + "\n" +
+///   <Scope> + "\n" +
+///   Hex(SHA256Hash(<CanonicalRequest>))
+///
+/// \param timestamp is an ISO8601 format timestamp
+/// \param is a scope string (date/region/service)
+/// \param canonical_req is a canonical request
+static result<std::string> get_string_to_sign(
+  const std::string& timestamp,
+  const std::string& scope,
+  const std::string& canonical_req) {
+    auto digest = sha_256(canonical_req);
+    if (!digest) {
+        vlog(s3_log.error, "string_to_sign error: {}", digest.error());
+        return digest.error();
+    }
+    std::string_view algorithm = "AWS4-HMAC-SHA256";
```

Comment:
constexpr?

Reply:
also this string is repeated a few times. why not extract it as a top level constexpr?